lib/inet-grouping.js

```diff
@@ -1,24 +1,122 @@
+ * @returns {Promise<string>}
+ */
+export const mapSubnetToInetGroup = async (pgClient, subnet, now = new Date()) => {
+  const { rows: [found] } = await pgClient.query(
+    'SELECT id FROM inet_groups WHERE subnet = $1',
+    [subnet]
+  )
+  if (found) return found.id
+
+  let remainingRetries = 5
+  for (;; remainingRetries--) {
+    const group = generateUniqueGroupId(now)
+    try {
+      // There is a race condition: between the time we looked up the existing group
+      // and the time we execute this query, a handler for a different request from
+      // the same subnet may have already defined the group for this subnet.
+      // Solution: when insert fails with a conflict in the group id, use that existing group
+      // instead of the value we generated ourselves.
+      //
+      // ON CONFLICT DO UPDATE is needed to tell PG to return the existing id on conflict
+      // ON CONFLICT DO NOTHING would not return the existing id on conflict
+      const { rows: [created] } = await pgClient.query(`
+        INSERT INTO inet_groups (id, subnet, created_at)
+        VALUES ($1, $2, $3)
+        ON CONFLICT (subnet) DO UPDATE SET id = inet_groups.id
+        RETURNING id
+       `, [
+        group,
+        subnet,
+        now
+      ])
+      return created.id
+    } catch (err) {
+      if (err.code === 23505 && err.constraint === 'inet_groups_pkey') {
+        // Retry with a different random id
+        if (remainingRetries > 0) continue
+        // Fail if we exhausted allowed attempts
+        throw new Error('Failed to generate a unique group id')
+      }
+      throw err
+    }
+  }
+}
+
+/**
+ * @param {Date} [now]
+ * @returns {Promise<string>}
+ */
+const generateUniqueGroupId = (now) => {
+  const data = new ArrayBuffer(8)
+
+  const prefix = new Uint8Array(data, 0, 2)
+
+  // To reduce the likelihood of generating the same group values for different subnets,
+  // and also to make it easier to diagnose invalid use of groups (e.g. after they should
+  // have been expired), let's prefix the group with a single byte derived from the number
+  // of hours elapsed since the Unix epoch.
+  // 251 is the largest prime number that fits into a single byte.
+  prefix[0] = now.getTime() / 3600_000 % 251
+
+  // We are running this code in more than one process. To reduce possible clashes, let's
+  // add another byte derived from the millisecond time.
+  prefix[1] = now.getTime() % 251
+
+  // Finally, add six random bytes
+  globalThis.crypto.getRandomValues(new Uint8Array(data, 2, 6))
+
+  return base64url.encode(new Uint8Array(data))
+}
+
+/**
+ * @param {import('node:http').IncomingMessage} req
+ * @returns {Promise<string>}
+ */
+export const mapRequestToSubnet = (req) => {
   let addr = req.socket.remoteAddress
 
   const flyClientAddr = req.headers['fly-client-ip']
   if (flyClientAddr) addr = flyClientAddr
 
+  // TODO accept the value in this header only when the request is coming from Cloudflare
```

Comment:
How can we solve this?

Reply:
See https://developers.cloudflare.com/support/troubleshooting/restoring-visitor-ips/restoring-original-visitor-ips/, which explains how to configure `mod_remoteip` module for Apache to correctly restore original visitor IP:

> Define trusted proxy addresses by creating /etc/apache2/conf-available/remoteip.conf by entering the following code and [ Cloudflare IPs](https://www.cloudflare.com/ips/):
>
> ```
> RemoteIPHeader CF-Connecting-IP
> RemoteIPTrustedProxy 192.0.2.1 (example IP address)
> RemoteIPTrustedProxy 192.0.2.2 (example IP address)
> (repeat for all Cloudflare IPs listed at https://www.cloudflare.com/ips/)
> ```

I improved the code comment to explain, see https://github.com/filecoin-station/spark-api/pull/104/commits/480d770f89cc3dc42bf159524dc985c925e5ce7f